libraries/ms-drivers/src/bts_7040_load_switch.c

```diff
@@ -0,0 +1,177 @@
+#include "bts_7040_load_switch.h"
+
+static void prv_measure_current(SoftTimerId timer_id, void *context) {
+  Bts7040Storage *storage = context;
+  bts_7040_get_measurement(storage, &storage->reading_out);
+
+  if (storage->callback != NULL) {
+    storage->callback(storage->reading_out, storage->callback_context);
+  }
+
+  soft_timer_start(storage->interval_us, &prv_measure_current, storage,
+                   &storage->measurement_timer_id);
+}
+
+// Init common elements of the load switch to STM32, PCA9539R implementations.
+static StatusCode prv_init_common(Bts7040Storage *storage) {
+  // make sure that if the user calls stop() it doesn't cancel some other timer
+  storage->measurement_timer_id = SOFT_TIMER_INVALID_TIMER;
+
+  // Timer for fault handling
+  storage->enable_pin.fault_timer_id = SOFT_TIMER_INVALID_TIMER;
+
+  storage->enable_pin.fault_in_progress = false;
+
+  // initialize the sense pin
+  GpioSettings sense_settings = {
+    .direction = GPIO_DIR_IN,
+    .alt_function = GPIO_ALTFN_ANALOG,
+  };
+  status_ok_or_return(gpio_init_pin(storage->sense_pin, &sense_settings));
+
+  // initialize the sense pin as ADC
+  AdcChannel sense_channel = NUM_ADC_CHANNELS;
+  adc_get_channel(*storage->sense_pin, &sense_channel);
+  adc_set_channel(sense_channel, true);
+  return STATUS_CODE_OK;
+}
+
+StatusCode bts_7040_init_stm32(Bts7040Storage *storage, Bts7040Stm32Settings *settings) {
+  storage->sense_pin = settings->sense_pin;
+
+  storage->enable_pin.enable_pin_stm32 = settings->enable_pin;
+  storage->enable_pin.enable_pin_pca9539r = NULL;
+
+  storage->enable_pin.pin_type = BTS7XXX_PIN_STM32;
+
+  storage->interval_us = settings->interval_us;
+  storage->callback = settings->callback;
+  storage->callback_context = settings->callback_context;
+  storage->fault_callback = settings->fault_callback;
+  storage->fault_callback_context = settings->fault_callback_context;
+
+  storage->resistor = settings->resistor;
+
+  storage->min_fault_voltage_mv = settings->min_fault_voltage_mv;
+  storage->max_fault_voltage_mv = settings->max_fault_voltage_mv;
+
+  // initialize the enable pin
+  GpioSettings enable_settings = {
+    .direction = GPIO_DIR_OUT,
+    .state = GPIO_STATE_LOW,
+    .resistor = GPIO_RES_NONE,
+    .alt_function = GPIO_ALTFN_NONE,
+  };
+
```

Comment:
I think a lot of this copy/pasting could instead be reused between the bts7200 and bts7040 drivers. Perhaps a `Bts7xxxStorage` struct is in order which contains all the common one-off fields (`resistor`, `min_fault_voltage_mv`, `enable_pin`, etc) between the two drivers? Then you could include a `Bts7xxxStorage` struct in `Bts7200Storage` and `Bts7040Storage` and factor out most of the copy/pasting here and elsewhere into non-public bts7xxx functions.

The only things that *really* have to be different between the two modules are the measurement reading and fault detection logic, I think that the rest of it could be factored out into `bts7xxx` by taking `Bts7xxxStorage *` pointers.